File: typescript/Behavioral/21-TemplateMethod/Report/CarpentryInspection.ts
import { JobInspectionReport } from "./JobInspectionReport";

export class CarpentryInspection extends JobInspectionReport {
  protected get tradeName(): string {
    return "CARPENTRY";
  }

  protected verifySafety(): void {
    console.log("║ SAFETY CHECK");
    console.log("║   ✓ Cable scan completed before cutting");
    console.log("║   ✓ Dust extraction in place");
    console.log("║   ✓ PPE worn throughout");
  }

  protected performInspection(): void {
    console.log("║ INSPECTION");
    console.log("║   → Level and plumb verification");
    console.log("║   → Fixing point load test");
    console.log("║   → Finish and edge quality review");
  }

  protected recordFindings(): void {
    console.log("║ FINDINGS");
    this.addFinding("All surfaces level within 2 mm tolerance");
    this.addFinding("Fixings rated for 45 kg — exceeds 25 kg requirement");
    this.addFinding("Edges sanded and sealed");
  }

  protected attachComplianceNote(): void {
    console.log("║ COMPLIANCE");
    console.log("║   Non-structural work — no building plan approval required");
  }
}

File: csharp/Behavioral/21-TemplateMethod/Report/CarpentryInspection.cs
namespace TemplateMethod;

public class CarpentryInspection : JobInspectionReport
{
    protected override string TradeName => "CARPENTRY";

    protected override void VerifySafety()
    {
        Console.WriteLine("║ SAFETY CHECK");
        Console.WriteLine("║   ✓ Cable scan completed before cutting");
        Console.WriteLine("║   ✓ Dust extraction in place");
        Console.WriteLine("║   ✓ PPE worn throughout");
    }

    protected override void PerformInspection()
    {
        Console.WriteLine("║ INSPECTION");
        Console.WriteLine("║   → Level and plumb verification");
        Console.WriteLine("║   → Fixing point load test");
        Console.WriteLine("║   → Finish and edge quality review");
    }

    protected override void RecordFindings()
    {
        Console.WriteLine("║ FINDINGS");
        AddFinding("All surfaces level within 2 mm tolerance");
        AddFinding("Fixings rated for 45 kg — exceeds 25 kg requirement");
        AddFinding("Edges sanded and sealed");
    }

    protected override void AttachComplianceNote()
    {
        Console.WriteLine("║ COMPLIANCE");
        Console.WriteLine("║   Non-structural work — no building plan approval required");
    }
}
